src/middleware/devtools.ts

```diff
@@ -61,107 +74,188 @@ export const devtools =
-          if (state !== (newState as any)[DEVTOOLS] && api.devtools) {
-            api.devtools.send(api.devtools.prefix + 'setState', api.getState())
+    const setStateFromDevtools: SetState<S> = (state, replace) => {
+      isRecording = false
+      if (!replace) {
+        set(state)
+      } else {
+        const currentState = get()
+        for (const k in currentState) {
+          if (typeof currentState[k] === 'function') {
+            ;(state as any)[k] = currentState[k]
           }
         }
+        set(state, true)
       }
-      options = typeof options === 'string' ? { name: options } : options
-      const connection = (api.devtools = extension.connect({ ...options }))
-      connection.prefix = options?.name ? `${options.name} > ` : ''
-      connection.subscribe((message: any) => {
-        if (message.type === 'ACTION' && message.payload) {
-          try {
-            api.setState(JSON.parse(message.payload))
-          } catch (e) {
-            console.error(
-              'please dispatch a serializable value that JSON.parse() support\n',
-              e
-            )
-          }
-        } else if (message.type === 'DISPATCH' && message.state) {
-          const jumpState =
-            message.payload.type === 'JUMP_TO_ACTION' ||
-            message.payload.type === 'JUMP_TO_STATE'
-          const newState = api.getState()
-          ;(newState as any)[DEVTOOLS] = JSON.parse(message.state)
-
-          if (!api.dispatch && !jumpState) {
-            api.setState(newState)
-          } else if (jumpState) {
-            api.setState((newState as any)[DEVTOOLS])
-          } else {
-            savedSetState(newState)
-          }
-        } else if (
-          message.type === 'DISPATCH' &&
-          message.payload?.type === 'COMMIT'
-        ) {
-          connection.init(api.getState())
-        } else if (
-          message.type === 'DISPATCH' &&
-          message.payload?.type === 'IMPORT_STATE'
-        ) {
-          const actions = message.payload.nextLiftedState?.actionsById
-          const computedStates =
-            message.payload.nextLiftedState?.computedStates || []
-
-          computedStates.forEach(
-            ({ state }: { state: PartialState<S> }, index: number) => {
-              const action = actions[index] || 'No action found'
-
-              if (index === 0) {
-                connection.init(state)
-              } else {
-                savedSetState(state)
-                connection.send(action, api.getState())
-              }
-            }
-          )
-        }
-      })
-      connection.init(initialState)
+      isRecording = true
     }
+
+    const initialState = fn(api.setState, get, api)
+    extension!.init(initialState)
+
+    extension!.subscribe((message: any) => {
+      switch (message.type) {
+        case 'ACTION':
+          if (!api.dispatchFromDevtools) return
+          if (typeof api.dispatch !== 'function') return
```

Comment:
The previous "dispatching" you're talking about was a bug, I explain it [here](https://github.com/pmndrs/zustand/discussions/673#discussioncomment-1676328). I've also mentioned it in short in the description of the PR:

> If the user dispatches `{ type: "INCREMENT" }` action from the devtools, instead of ignoring it (or dispatching if it's a redux store) we used to update the state with it, so the new state would become `{ ...state, type: "INCREMENT" }` this is fixed.